### ISSUE TITLE
Allow Team Library access from Dev Mode devices

### ISSUE THREAD

**Issue by griemenschneider** [flowforge/flowforge#2294]
### Description

As a FF developer that exclusively uses devices I would like my team to utilize the flows library functionality offered by FlowForge.

Currently, that option is missing from the menu:

![image](https://github.com/flowforge/flowforge/assets/135891848/8765e982-b403-462f-9dde-42f9b1b82a66)

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

**Comment by hardillb** [comment]
Notes on current situation:

Libraries are provided by the storage plugin, currently the device-agent uses the standard filesystem storage plugin so has no access to the shared team library.

**Comment by griemenschneider** [comment]
@hardillb Can this be resolved by adding "@flowforge/nr-storage" to the device's package.json or will it conflict with the device-agent filesystem plugin?

**Comment by hardillb** [comment]
@griemenschneider no, this would require a new storage plugin to be written that mixes loading flows from local storage and libraries remotely.

**Comment by hardillb** [comment]
There are also security considerations required here with token scopes for devices not having access to the library endpoints.

**Comment by knolleary** [comment]
Just to clarify what is needed here.

The Shared Library is provided by a *library plugin* and not the storage module. It happens to be the case that both of these things are currently packaged together in the `nr-launcher` module that is only used in the hosted instances.

The required tasks are:

 - Make the existing library plugin available to the device agent
    - either done by pulling the plugin out of nr-launcher and into its own npm package (which goes against out attempt to consolidate the modules we publish)
    - or use some sort of git submodule approach to pull the files into the device agent when it is build and published.
 - Add the additional configuration needed to the device agent settings file
 - Add the necessary scopes to the device agent token

The main issue is the security side. We minimise the scope of what a device token is able to do on the platform as the devices exist outside the platform security boundary. If the device is given full read/write access to the library the risk is anyone with physical access to the device will be able to extract the token and have that same access. That *might* be an acceptable risk for some users, but not necessarily all. That is the piece we need to make sure we have a good position on.

**Comment by griemenschneider** [comment]
Regarding the security side I would add that r/w access to the library should only be granted to device in Dev Mode - devices outside of Dev Mode really have no business accessing the library. Since typical use cases involve devices being outside of Dev Mode this should limit the risk. In addition, if physical access to the device can be achieved then physical access to the code on that device has been achieved. So where is the boundary between security of the device-side code and the library-side code? As long as credentials are not stored in the library and changes to the library are logged I would argue the risks are further minimized

**Comment by knolleary** [comment]
Closed as one of the PRs has been merged, but definitely on track for next week's 1.10 release.

**Comment by Steve-Mcl** [comment]
verified on staging:

- [x] download flow from team lib to remote device
- [x] upload flow from team lib to remote device
- [x] download function code from team lib to remote device
- [x] upload function code from team lib to remote device
- [x] team lib correctly displays uploaded flows
- [x] team lib correctly displays uploaded function code


### team library demo of flow import/export



https://github.com/flowforge/flowforge/assets/44235289/364d670a-21b5-4c06-bced-7c5ee134ec51



### team library demo of function code import/export


https://github.com/flowforge/flowforge/assets/44235289/efa04ba4-3a98-48ba-aabd-270d6d714f22




### team library shows uploaded flow

![Image](https://github.com/flowforge/flowforge/assets/44235289/7da68d3f-0c91-4db3-a533-ad71b088c2b7)

### team library shows uploaded function code

![Image](https://github.com/flowforge/flowforge/assets/44235289/950e609c-9a96-4d33-ba25-ed441090b188)
